lib/auth/auth_with_roles.go

```diff
@@ -610,18 +610,167 @@ func (a *ServerWithRoles) GenerateHostCerts(ctx context.Context, req *proto.Host
 	if a.context.User.GetName() != HostFQDN(req.HostID, clusterName) {
 		return nil, trace.AccessDenied("username mismatch %q and %q", a.context.User.GetName(), HostFQDN(req.HostID, clusterName))
 	}
+
 	existingRoles, err := types.NewTeleportRoles(a.context.User.GetRoles())
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
-	// prohibit privilege escalations through role changes
-	if !existingRoles.Equals(types.SystemRoles{req.Role}) {
+	if req.Role == types.RoleInstance {
```

Comment:
There are a lot of different code paths through here with all the looping and branching - I would be more comfortable with some tests to exercise this functionality.

Reply:
Added unit test for happy path, and for attempting to request system roles without authorization.